### PR TITLE
Add PCP netatop BPF PMDA to extract per process network statistic.

### PR DESCRIPTION
**Pull request by stanfordcox**
Add a bpf per process network metric PMDA which exposes per process network metrics using netatop-bpf.c. Add display of metrics to atop. Add test for network metrics.

### CODE REVIEW

File: qa/1264 [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+_prepare_pmda bpf
+trap "_cleanup; exit \$status" 0 1 2 3 15
+_stop_auto_restart pmcd
+
+# real QA test starts here
+
+_pmdabpf_install $tmp.conf
+sudo ln -s  /usr/local/bpf.log /usr/local//var/log/pcp//pmcd/
```
Comment (natoscott) [inline qa/1264:53]: I think this may be a leftover from local work / debugging?

File: src/pmdas/bpf/modules/netatop.c [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+static unsigned int netatop_indom_count(void)
+{
+    return INDOM_COUNT;
+}
+
+static void netatop_set_indom_serial(unsigned int local_indom_id, unsigned int global_id)
+{
+    pmNotifyErr(LOG_INFO, "%s", __FUNCTION__);
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.c:121]: Here and in several other places we have unguarded diagnostics that may be filling the PMDA bpf.log file during operation.  Could we add guards for all/most of these such as:

```
if (pmDebugOptions.appl0 || pmDebugOptions.desperate)
    pmNotifyErr(LOG_INFO, "%s", __FUNCTION__);
```

The full list of debug flags is available running: 'pmdbg -l'.

File: src/pmdas/bpf/modules/netatop.c [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    }
+
+    return 0;
+}
+
+
+static void netatop_register(unsigned int cluster_id, pmdaMetric *metrics, pmdaIndom *indoms)
+{
+    pmNotifyErr(LOG_INFO, "%s Adding netatop metrics", __FUNCTION__);
+    /* bpf.netatop.tcpsndpacks */
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.c:140]: The metric name in the comment (here and below) doesn't match the namespace used - I guess this changed over time?

File: src/pmdas/bpf/modules/netatop.c [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+enum metric_name {
+  TCPSNDPACKS,
+  TCPSNDBYTES,
+  TCPRCVPACKS,
+  TCPRCVBYTES,
+  UDPSNDPACKS,
+  UDPSNDBYTES,
+  UDPRCVPACKS,
+  UDPRCVBYTES };
+enum metric_indom { NETATOP_INDOM };
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.c:58]: (consistency) unusual coding style for this enum metric_name.

File: src/pmdas/bpf/modules/netatop.c [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+static struct env {
+    int process_count;
+} env = {
+    .process_count = 20,
+};
+
+int tgid_map_fd;
+int nr_cpus;
+
+static pmdaInstid *netatop_instances = NULL;
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.c:40]: No need to assign to NULL here, its mandatory for the compiler to do this for globals.

File: src/pmdas/bpf/modules/netatop.c [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+#include <signal.h>
+#include <string.h>
+#include <errno.h>
+#include <dirent.h>
+#include <sys/resource.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include <pcp/pmda.h>
+#include <sys/queue.h>
+#include <pcp/pmwebapi.h>
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.c:20]: should this be <pcp/pmapi.h>?  (not sure - do we use something from webapi here?)
Either way, pmapi.h is included via pmwebapi.h and it includes the headers from line 9 thru 14 I think, so those could be dropped.

File: src/pmdas/bpf/modules/netatop.c [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+#include <string.h>
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.c:21]: string.h is included twice too.

File: src/pmdas/bpf/modules/netatop.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+};
+
+extern struct netatop_bpf *skel;
+extern int semid;
+extern int tgid_map_fd;
+extern int tid_map_fd;
+extern int nr_cpus;
+
+#define	NUMCLIENTS	(semctl(semid, 1, GETVAL, 0))
+#define SEMAKEY         1541962
```
Comment (natoscott) [inline src/pmdas/bpf/modules/netatop.h:45]: Are NUMCLIENTS and SEMAKEY used?  I may be missing it, but if not let's drop these two lines.

**Review by natoscott — COMMENTED** [review]
@stanfordcox looking great Stan!  Some small improvement suggestions here, but it looks pretty close to done now.

File: src/pmdas/bpf/modules/netatop.c (hunk 6 above)
Comment (stanfordcox) [reply src/pmdas/bpf/modules/netatop.c]: yep, most of the bpf/modules routines include pmwebapi.h for the pmIniFileLookup dcl so I added it back.  Without it rawhide does not build
netatop.c: In function ‘netatop_init’:
netatop.c:271:16: error: implicit declaration of function ‘pmIniFileLookup’ [-Wimplicit-function-declaration]
  271 |     if ((val = pmIniFileLookup(cfg, module_name, "process_count")))

**Pull request merged**